alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,180 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Error;
+use std::os::raw::c_void;
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+use log::warn;
+
+use mio_extras::channel::{channel, Receiver};
+
+use winapi::shared::ntdef::{BOOLEAN, HANDLE, PVOID};
+use winapi::um::winbase::{RegisterWaitForSingleObject, UnregisterWait, INFINITE};
+use winapi::um::winnt::{WT_EXECUTEINWAITTHREAD, WT_EXECUTEONLYONCE};
+
+use crate::tty::ChildEvent;
+
+/// WinAPI callback for `HandleWaitSignal`, unpacks Rust closure reference and calls it
+extern "system" fn child_exit_callback<F>(ctx: PVOID, timed_out: BOOLEAN)
+where
+    F: FnOnce() + Send,
+{
+    if timed_out != 0 {
+        return;
+    }
+
+    let callback: Box<F> = unsafe { Box::from_raw(ctx as *mut F) };
+    callback();
+}
+
+/// Represents closure attached to Win32 handle wait signal
+///
+/// This allows to fire a Rust callback when subprocess exits
+pub(crate) struct HandleWaitSignal {
+    wait_handle: AtomicPtr<c_void>,
+}
+
+impl HandleWaitSignal {
+    /// Registers an asynchronous closure to call when `child_handle` wait signals.
```

Comment:
> when `child_handle` wait signals.

What do you mean by this?

Reply:
I agree it is phrased in a confusing manner. The struct is just a thin wrapper between WinAPI and Rust closures. The wording is paraphrase of docs of [RegisterWaitForSingleObject] where it specifically says that callback is called when:

> The specified object is in the signaled state.

The function works with any of [synchronisation objects], although I only use it for synchronising on Process exiting.

I guess if you don't care about the that level of details, this could be changed to
```suggestion
    /// Registers an asynchronous closure to call when process under `child_handle` exits.
```

If you want this to be more precise, I guess the variable should just be called `handle` and comment could be more along the line of
```suggestion
    /// Registers an asynchronous closure to call when WinAPI Synchronization Object under `handle` sends a signal.
```

Let me know which one seems more fitting.

[RegisterWaitForSingleObject]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-registerwaitforsingleobject?redirectedfrom=MSDN 
[synchronisation objects]: https://docs.microsoft.com/en-us/windows/win32/sync/synchronization-objects

Reply:
I like the first one. Describe what we do, not what we might be doing.